File: apps/clojure/part1/spec/composition_spec.clj
(ns composition-spec
  (:require [speclj.core :refer :all]
            [composition :refer :all]))

(describe "関数合成と高階関数"

  (context "関数合成の基本 (comp)"
    (it "calculate-final-price は割引、税金、丸めを適用する"
      ;; 1000円 -> 20%割引で800円 -> 10%税込で880円
      (should= 880 (calculate-final-price 1000)))

    (it "add-tax は税金を追加する"
      (should= 1100.0 (add-tax 0.1 1000)))

    (it "apply-discount-rate は割引を適用する"
      (should= 800.0 (apply-discount-rate 0.2 1000)))

    (it "round-to-yen は円単位に丸める"
      (should= 100 (round-to-yen 99.5))
      (should= 99 (round-to-yen 99.4))))

  (context "部分適用 (partial)"
    (it "say-hello は Hello で挨拶する"
      (should= "Hello, 田中!" (say-hello "田中")))

    (it "say-goodbye は Goodbye で挨拶する"
      (should= "Goodbye, 鈴木!" (say-goodbye "鈴木")))

    (it "send-notification は部分適用されたメールを送信する"
      (let [email (send-notification "メッセージ本文")]
        (should= "system@example.com" (:from email))
        (should= "user@example.com" (:to email))
        (should= "通知" (:subject email))
        (should= "メッセージ本文" (:body email)))))

  (context "juxt - 複数の関数を並列適用"
    (it "get-stats は数値リストの統計情報を返す"
      (let [[first-val last-val cnt min-val max-val] (get-stats [3 1 4 1 5 9 2 6])]
        (should= 3 first-val)
        (should= 6 last-val)
        (should= 8 cnt)
        (should= 1 min-val)
        (should= 9 max-val)))

    (it "analyze-person は人物情報を分析する"
      (should= ["田中" 25 :adult] (analyze-person {:name "田中" :age 25}))
      (should= ["鈴木" 15 :minor] (analyze-person {:name "鈴木" :age 15}))))

  (context "高階関数によるデータ処理"
    (it "process-with-logging はログを出力しながら処理を実行する"
      (let [add-with-log (process-with-logging +)]
        (should= 5 (add-with-log 2 3))))

    (it "retry は失敗時にリトライする"
      (let [counter (atom 0)
            flaky-fn (fn []
                       (swap! counter inc)
                       (if (< @counter 3)
                         (throw (Exception. "一時的なエラー"))
                         "成功"))
            retry-fn (retry flaky-fn 5)]
        (should= "成功" (retry-fn))
        (should= 3 @counter))))

  (context "パイプライン処理"
    (it "pipeline は関数を順次適用する"
      (let [double-then-add10 (pipeline #(* % 2) #(+ % 10))]
        (should= 30 (double-then-add10 10))))

    (it "process-order-pipeline は注文を処理する"
      (let [order {:items [{:price 1000 :quantity 2}
                           {:price 500 :quantity 3}]
                   :customer {:membership :gold}}
            result (process-order-pipeline order)]
        ;; 合計: 2000 + 1500 = 3500
        ;; 割引: 3500 * 0.9 = 3150
        ;; 送料: 3150 >= 5000 ? 0 : 500 => 500
        ;; 最終: 3150 + 500 = 3650
        (should= 3650.0 (:total result))
        (should= 500 (:shipping result))))

    (it "validate-order は空の注文でエラーを投げる"
      (should-throw clojure.lang.ExceptionInfo
                    (validate-order {:items []}))))

  (context "関数合成によるバリデーション"
    (it "validate-quantity は有効な数量を検証する"
      (should (:valid (validate-quantity 50)))
      (should-not (:valid (validate-quantity -1)))
      (should-not (:valid (validate-quantity 100)))
      (should-not (:valid (validate-quantity 1.5))))

    (it "validator は述語に基づいてバリデーションする"
      (let [v (validator even? "偶数である必要があります")]
        (should (:valid (v 2)))
        (should-not (:valid (v 3)))
        (should= "偶数である必要があります" (:error (v 3))))))

  (context "関数の変換"
    (it "flip は引数の順序を反転する"
      (should= 2 ((flip -) 3 5)))

    (it "complement-fn は述語の結果を反転する"
      (let [not-even? (complement-fn even?)]
        (should (not-even? 3))
        (should-not (not-even? 2))))

    (it "constantly-fn は常に同じ値を返す"
      (let [always-42 (constantly-fn 42)]
        (should= 42 (always-42))
        (should= 42 (always-42 1 2 3))))

    (it "curry は2引数関数をカリー化する"
      (let [curried-add (curry +)
            add5 (curried-add 5)]
        (should= 8 (add5 3))))

    (it "uncurry はカリー化された関数を元に戻す"
      (let [curried (fn [a] (fn [b] (+ a b)))
            uncurried (uncurry curried)]
        (should= 8 (uncurried 5 3)))))

  (context "関数合成のパターン"
    (it "valid-age? は有効な年齢をチェックする"
      (should (valid-age? 25))
      (should (valid-age? 1))
      (should (valid-age? 150))
      (should-not (valid-age? -1))
      (should-not (valid-age? 151))
      (should-not (valid-age? 25.5)))

    (it "premium-customer? はプレミアム顧客をチェックする"
      (should (premium-customer? {:membership :gold :purchase-count 0 :total-spent 0}))
      (should (premium-customer? {:membership :bronze :purchase-count 100 :total-spent 0}))
      (should (premium-customer? {:membership :bronze :purchase-count 0 :total-spent 100000}))
      (should-not (premium-customer? {:membership :bronze
                                       :purchase-count 10
                                       :total-spent 1000})))))

(run-specs)

File: apps/scala/part1/src/test/scala/CompositionSpec.scala
package composition

import org.scalatest.flatspec.AnyFlatSpec
import org.scalatest.matchers.should.Matchers

class CompositionSpec extends AnyFlatSpec with Matchers:

  import Composition.*

  // =============================================================================
  // 関数合成の基本 (andThen, compose)
  // =============================================================================

  "calculateFinalPrice" should "割引、税金、丸めを適用する" in {
    // 1000円 -> 20%割引で800円 -> 10%税込で880円
    calculateFinalPrice(1000) shouldBe 880
  }

  "addTax" should "税金を追加する" in {
    addTax(0.1)(1000) shouldBe 1100.0
  }

  "applyDiscountRate" should "割引を適用する" in {
    applyDiscountRate(0.2)(1000) shouldBe 800.0
  }

  "roundToYen" should "円単位に丸める" in {
    roundToYen(99.5) shouldBe 100
    roundToYen(99.4) shouldBe 99
  }

  "calculateFinalPriceCompose" should "composeでも同じ結果を返す" in {
    calculateFinalPriceCompose(1000) shouldBe 880
  }

  // =============================================================================
  // 部分適用 (カリー化)
  // =============================================================================

  "sayHello" should "Hello で挨拶する" in {
    sayHello("田中") shouldBe "Hello, 田中!"
  }

  "sayGoodbye" should "Goodbye で挨拶する" in {
    sayGoodbye("鈴木") shouldBe "Goodbye, 鈴木!"
  }

  "sendNotification" should "部分適用されたメールを送信する" in {
    val email = sendNotification("メッセージ本文")
    email.from shouldBe "system@example.com"
    email.to shouldBe "user@example.com"
    email.subject shouldBe "通知"
    email.body shouldBe "メッセージ本文"
  }

  // =============================================================================
  // 複数の関数を並列適用
  // =============================================================================

  "getStats" should "数値リストの統計情報を返す" in {
    val (firstVal, lastVal, cnt, minVal, maxVal) = getStats(List(3, 1, 4, 1, 5, 9, 2, 6))
    firstVal shouldBe 3
    lastVal shouldBe 6
    cnt shouldBe 8
    minVal shouldBe 1
    maxVal shouldBe 9
  }

  "analyzePerson" should "人物情報を分析する" in {
    analyzePerson(Map("name" -> "田中", "age" -> 25)) shouldBe PersonAnalysis("田中", 25, "adult")
    analyzePerson(Map("name" -> "鈴木", "age" -> 15)) shouldBe PersonAnalysis("鈴木", 15, "minor")
  }

  // =============================================================================
  // 高階関数によるデータ処理
  // =============================================================================

  "processWithLogging" should "ログを出力しながら処理を実行する" in {
    val doubleWithLog = processWithLogging[Int, Int](_ * 2)
    doubleWithLog(5) shouldBe 10
  }

  "retry" should "失敗時にリトライする" in {
    var counter = 0
    val flakyFn: Int => String = _ => {
      counter += 1
      if counter < 3 then throw new RuntimeException("一時的なエラー")
      else "成功"
    }
    val retryFn = retry(flakyFn, 5)
    retryFn(0) shouldBe "成功"
    counter shouldBe 3
  }

  // =============================================================================
  // パイプライン処理
  // =============================================================================

  "pipeline" should "関数を順次適用する" in {
    val doubleThenAdd10 = pipeline[Int](_ * 2, _ + 10)
    doubleThenAdd10(10) shouldBe 30
  }

  "processOrderPipeline" should "注文を処理する" in {
    val order = Order(
      items = List(OrderItem(1000, 2), OrderItem(500, 3)),
      customer = Customer("gold")
    )
    val result = processOrderPipeline(order)
    // 合計: 2000 + 1500 = 3500
    // 割引: 3500 * 0.9 = 3150
    // 送料: 3150 >= 5000 ? 0 : 500 => 500
    // 最終: 3150 + 500 = 3650
    result.total shouldBe 3650.0
    result.shipping shouldBe 500
  }

  "validateOrder" should "空の注文でエラーを投げる" in {
    an[IllegalArgumentException] should be thrownBy {
      validateOrder(Order(items = List.empty, customer = Customer("regular")))
    }
  }

  // =============================================================================
  // 関数合成によるバリデーション
  // =============================================================================

  "validateQuantity" should "有効な数量を検証する" in {
    validateQuantity(50).valid shouldBe true
    validateQuantity(-1).valid shouldBe false
    validateQuantity(100).valid shouldBe false
  }

  "validator" should "述語に基づいてバリデーションする" in {
    val v = validator[Int](_ % 2 == 0, "偶数である必要があります")
    v(2).valid shouldBe true
    v(3).valid shouldBe false
    v(3).error shouldBe Some("偶数である必要があります")
  }

  // =============================================================================
  // 関数の変換
  // =============================================================================

  "flip" should "引数の順序を反転する" in {
    val subtract = (a: Int, b: Int) => a - b
    flip(subtract)(3, 5) shouldBe 2  // 5 - 3 = 2
  }

  "complementFn" should "述語の結果を反転する" in {
    val isEven = (x: Int) => x % 2 == 0
    val notEven = complementFn(isEven)
    notEven(3) shouldBe true
    notEven(2) shouldBe false
  }

  "constantlyFn" should "常に同じ値を返す" in {
    val always42 = constantlyFn[Any, Int](42)
    always42(()) shouldBe 42
    always42("anything") shouldBe 42
  }

  "curry" should "2引数関数をカリー化する" in {
    val add = (a: Int, b: Int) => a + b
    val curriedAdd = curry(add)
    val add5 = curriedAdd(5)
    add5(3) shouldBe 8
  }

  "uncurry" should "カリー化された関数を元に戻す" in {
    val curried = (a: Int) => (b: Int) => a + b
    val uncurried = uncurry(curried)
    uncurried(5, 3) shouldBe 8
  }

  // =============================================================================
  // 関数合成のパターン
  // =============================================================================

  "validAge" should "有効な年齢をチェックする" in {
    validAge(25) shouldBe true
    validAge(1) shouldBe true
    validAge(150) shouldBe true
    validAge(-1) shouldBe false
    validAge(151) shouldBe false
  }

  "premiumCustomer" should "プレミアム顧客をチェックする" in {
    premiumCustomer(CustomerInfo("gold", 0, 0)) shouldBe true
    premiumCustomer(CustomerInfo("bronze", 100, 0)) shouldBe true
    premiumCustomer(CustomerInfo("bronze", 0, 100000)) shouldBe true
    premiumCustomer(CustomerInfo("bronze", 10, 1000)) shouldBe false
  }
